### PR TITLE
Metric Exemplars SDK Specification

### LINKED ISSUE #1797: Define Exemplar requirements in the Metrics SDK spec
**What are you trying to achieve?**

The metrics data model specification has covered Exemplar [here](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/datamodel.md#exemplars).

The goal is to have the SDK specification supporting exemplars.

Related to #1260.

### PR DESCRIPTION
**Pull request by jsuereth**
Fixes #1797

## Changes

Add Exemplar specification and customization to the metrics specification.

Related issues #1260 #617

Related [oteps](https://github.com/open-telemetry/oteps) [#113](https://github.com/open-telemetry/oteps/blob/main/text/metrics/0113-exemplars.md)

## Notes

- This exposes two extensible interfaces.  It's fine to drop these if they are contentious. The key important piece we must keep is the ability to turn OFF exemplars and the ability to limit exemplars to only measurements with associated sampled spans on context.
- This diverges from the Exemplar OTEP in that it is less restrictive on behavior of the exemplar reservoir (currently).  Additionally the aggregators available when the exemplar OTEP was written are not yet specified. We can refine this in a follow on PR as necessary.  We should align any sampling decisions w/ the future of tracing sampling.
- There are still some TODOs for discussion.

### CODE REVIEW

File: specification/metrics/sdk.md [hunk 1]
```diff
@@ -273,6 +273,75 @@ in the SDK:
...
+- value
+- `Attributes`
+- `Context`
+- timestamp
+
+The "offer" method SHOULD have the ability to pull associated trace and span
+information without needing to record full context.
+
+The "offer" method does not need to store all measurements it is given and
+MAY further sample beyond the `ExemplarFilter`.
```
Comment (victlu) [inline specification/metrics/sdk.md:330]: Can you clarify this a bit more?  I don't understand how it `MAY further sample beyond...`.

File: specification/metrics/sdk.md [hunk 2]
```diff
@@ -106,6 +106,75 @@ in the SDK:
...
+- The `value` that was recorded.
+- The `time` the measurement was seen.
+- The set of [Attributes](../common/common.md#attributes) associated with the measurement not already included in a metric data point.
+- The associated [trace id and span id](../trace/api.md#retrieving-the-traceid-and-spanid) of the active parent [Span within Context](../trace/api.md#determining-the-parent-span-from-a-context) of the measurement.
+
+A Metric SDK MUST provide a mechanism to sample `Exemplar`s from measurements.
+
+A Metric SDK MUST allow `Exemplar` sampling to be disabled.  In this instance the SDK SHOULD not have memory overhead related to exemplar sampling.
+
+A Metric SDK MUST sample `Exemplar`s only from measurements within the context of a sampled trace BY DEFAULT.
```
[the remark this answers is not in the document]
Comment (jsuereth) [reply specification/metrics/sdk.md:290]: The default behavior is you don't get exemplars if you aren't sampling traces, yes.  You would need to configure (via environment variable) exemplars.

File: specification/metrics/sdk.md [hunk 3]
```diff
@@ -106,6 +106,75 @@ in the SDK:
...
+
+<!-- Note:  I'm fine dropping these interfaces for the first release. -->
+A Metric SDK SHOULD provide extensible hooks for Exemplar sampling, specifically:
+
+- `ExemplarFilter`: filter which measurements can become exemplars
+- `ExemplarReservoir`: determine how to store exemplars.
+
+### Exemplar Filter
+
+The `ExemplarFilter` interface MUST provide a method to determine if a
```
[the remark this answers is not in the document]
Comment (jsuereth) [reply specification/metrics/sdk.md:304]: The three examples are listed below in the config section:  NONE, ALL and WITH_SAMPLED_TRACE.  Should I also list those here?

**Review by jmacd — COMMENTED** [review]
LGTM

File: specification/metrics/sdk.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (weyert) [reply specification/metrics/sdk.md:304]: Yeah, maybe adding a callout to the default filters listed below might be useful?

File: specification/metrics/sdk.md [hunk 4]
```diff
@@ -273,6 +273,75 @@ in the SDK:
...
+
+<!-- TODO: The wording of this is completely dependend on the Aggreagor specification. -->
+A Metric SDK MUST allow exemplar sampling to leverage the configuration of a metric aggregator.  For example,
+Exemplar sampling of histograms should be able to leverage bucket boundaries.
+
+<!-- Note:  I'm fine dropping these interfaces for the first release. -->
+A Metric SDK SHOULD provide extensible hooks for Exemplar sampling, specifically:
+
+- `ExemplarFilter`: filter which measurements can become exemplars
+- `ExemplarReservoir`: determine how to store exemplars.
```
[the remark this answers is not in the document]
Comment (jsuereth) [reply specification/metrics/sdk.md:300]: The basic idea of allocating a reservoir and storing measurements within it fits I think.   I'm also suggesting reservoir sampling for the only required implementation.

File: specification/metrics/sdk.md [hunk 5]
```diff
@@ -273,6 +273,75 @@ in the SDK:
...
+
+The "offer" method SHOULD accept measurements, including:
+
+- value
+- `Attributes`
+- `Context`
+- timestamp
+
+The "offer" method SHOULD have the ability to pull associated trace and span
+information without needing to record full context.
```
[the remark this answers is not in the document]
Comment (jsuereth) [reply specification/metrics/sdk.md:327]: yes.  Saving the entire context object should not be required.

File: specification/metrics/sdk.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (jsuereth) [reply specification/metrics/sdk.md:330]: They have different levels of understanding of the problem.  Basically:

- ExemplarFilter allows you to limit exemplars to ONLY traces or other context related-filters.  It is intended to be mostly configuration-based sampling.
- ExemplarReservoir does code-based sampling, e.g. sampling latest-measurement-per-histogram-bucket.   This is NOT intended for configuration but instead code/SPI-provided solutions.

File: specification/metrics/sdk.md [hunk 6]
```diff
@@ -273,6 +273,75 @@ in the SDK:
...
+
+The "offer" method does not need to store all measurements it is given and
+MAY further sample beyond the `ExemplarFilter`.
+
+The "collect" method MUST return accumulated `Exemplar`s.  
+
+These `Exemplar`s MUST have their attributes filtered by an aggregator
```
[the remark this answers is not in the document]
Comment (jsuereth) [reply specification/metrics/sdk.md:334]: - No this doesn't have to be part of the specification.
- Yes, assuming we keep this.

**Comment by jmacd** [comment]
FYI I have been working on trace sampling lately, and there is a lot of alignment between that work and this work. I have made claims that we can use weighted sampling to implement a hard rate limit on spans inside a tail sampler, and by extension I claim to know how to do the same for histogram exemplars. For a histogram aggregator outputting high-resolution histogram buckets you might not want one exemplar per bucket, and with weighted sampling we can produce a fixed number (i.e., using a reservoir) of exemplars with an equal expected number of spans per bucket.

This simple approach uses two passes, documented in the [README](https://github.com/lightstep/varopt#usage-inverse-probability-weights) of the https://github.com/lightstep/varopt repository. The more complicated approach can do this approximately in a single pass, which I also discussed briefly in today's Sampling SIG. See this [draft](https://github.com/lightstep/varopt/pull/18). I will elaborate on this in a prototype exemplar reservoir tail sampler based on Varopt.

**Comment by jsuereth** [comment]
@jmacd I agree I think we can do way better than status quo, especially for high-fidelity histograms.  I re-tweaked the wording to leave room for an exemplar reservoir sampling specification when those land.

**Review by jmacd — APPROVED** [review]

**Pull request merged**